### PR TITLE
Use SQPOLL mode & Kernel 5.11

### PR DESCRIPTION
**Pull request by 1Jo1**
Motivation:

if SQPOLL mode is enabled, 0-syscall submissionqueue, a kernel thread is created to perform submission queue polling which means that no submission is required, however the cpu usage is higher

Modifications:

-when the kernel thread is idle, we have to wake up 
if the ring tail is changed
-no io_uring_enter(submission) call anymore

Result:

better performance but the cpu usage is higher

### CODE REVIEW

**Comment by 1Jo1** [comment]
you cannot open/close descriptors in 5.9 as registration of files is required when SQPOLL is enabled but not 5.10, that's why I use 5.10 
if you want to test it you should use kernel (io_uring-5.10)[https://git.kernel.org/pub/scm/linux/kernel/git/axboe/linux-block.git/log/?h=for-5.10/io_uring]

**Comment by 1Jo1** [comment]
BTW I had a performance boost of 15% in echo benchmark tests when SQPOLL is active :fire:

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java [hunk 1]
```diff
@@ -136,11 +136,11 @@ private void setData(long sqe, byte op, int pollMask, int fd, long bufferAddress
...
 
-        // TODO: Make it configurable if we should use this flag or not.
-        PlatformDependent.putByte(sqe + SQE_FLAGS_FIELD,
-                // Workaround for a kernel bug:
-                // See https://lore.kernel.org/io-uring/6428c1ee0234105d18c5e3e88aa00c57@nickhill.org/T/#t
-                (byte) (op != Native.IORING_OP_WRITEV ? Native.IOSQE_ASYNC : 0));
+
+
+
+        // just a workaround, it seems to fail with
```
Comment (1Jo1) [inline transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java:142]: just temporary

**Comment by 1Jo1** [comment]
more details about SQPOLL in case if you interested [io_uring mailing list](https://lore.kernel.org/io-uring/CAAss7+pjbh2puVsQTOt7ymKSmbruBZbaOvB8tqfw0z-cMuhJYg@mail.gmail.com/T)

**Comment by 1Jo1** [comment]
I think it would be make sense to make  `sq_thread_cpu` configurable which is useful for servers with multiple cpus and `sq_thread_idle` default idle time is 1 Sec
[io_uring_params](https://github.com/axboe/liburing/blob/master/src/include/liburing/io_uring.h#L243)

**Comment by hyperxpro** [comment]
Would you mind sharing benchmark code?

**Comment by normanmaurer** [comment]
@hyperxpro we will share some more numbers etc soonish.

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java [hunk 2]
```diff
@@ -325,10 +326,19 @@ private int flushSqe() {
...
+            if (submitted > 0) {
+                int ret = Native.ioUringEnter(ringFd, submitted, 0, 0);
+                if (ret < 0) {
+                    throw new RuntimeException("ioUringEnter syscall");
+                }
+            }
+        } else {
+
+            //release barrier kFlagAddress
+            if((submitted > 0) && (PlatformDependent.getIntVolatile(kFlagsAddress) & Native.IORING_NEED_WAKEUP) != 0) {
```
Comment (njhill) [inline transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java:339]: nit: could move `submitted > 0` check outside

**Review by njhill — COMMENTED** [review]
@1Jo1 nice! I guess in this mode we may want to update the tail pointer after every SQE is added, i.e. better to not batch

**Comment by 1Jo1** [comment]
> @1Jo1 nice! I guess in this mode we may want to update the tail pointer after every SQE is added, i.e. better to not batch

@njhill yeah that's true, it's more efficient especially if you want to reduce the latency, I'm gonna test both cases in terms of throughput and latency

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java (hunk 1 above)
Comment (normanmaurer) [reply transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java:139]: Remove comment

**Comment by 1Jo1** [comment]
I forgot to mention that SQPOLL in 5.10 root permission is not required anymore, you just need CAP_SYS_NICE privilege

**Comment by normanmaurer** [comment]
Let's hold back on this till 5.10 was cut.

**Comment by normanmaurer** [comment]
BTW how we handle writev currently will not work with sqpoll as we can only re-use once the completion event comes back (no stable submission on io_uring_enter(...)).

**Comment by 1Jo1** [comment]
@normanmaurer I'm not sure what you mean here, I think it's stable, we do not call io_uring_enter, after you have changed the tail, only then the kernel will see what has been submitted

**Comment by normanmaurer** [comment]
@1Jo1 as we aren't most of the times call `io_uring_enter` we can not assume we are able to reuse the `IovArray`(as this is only save once we call `io_uring_enter`). We only know for sure once we received the completion event. I have a possible fix for this in the works. Gimme an hour

**Comment by 1Jo1** [comment]
@normanmaurer sorry I don't get it.. IovArray should be reuseable as `submissionCallback` will be called after we change the tail

**Comment by 1Jo1** [comment]
btw I try to rebase

**Comment by normanmaurer** [comment]
@1Jo1 thats not correct... we can only re-use once we know the kernel consumed it or we did call `io_uring_enter(...)`. As we not call `io_uring_enter(...)` most of the times here we need to be sure the kernel processed it. Even after we changed the tail the kernel may not consume it for some time so its not safe to reuse until we are sure.

**Comment by 1Jo1** [comment]
> @1Jo1 thats not correct... we can only re-use once we know the kernel consumed it or we did call `io_uring_enter(...)`. As we not call `io_uring_enter(...)` most of the times here we need to be sure the kernel processed it. Even after we changed the tail the kernel may not consume it for some time so its not safe to reuse until we are sure.

yes that's true, but that's not related to the SQPOLL implementation, it's general problem, even if we call `io_uring_enter` we don't know when the writev event will be processed, I agree as well we should reused when we will get the completion writev event

**Comment by normanmaurer** [comment]
@1Jo1 thats incorrect... as soon as you call `io_uring_enter` you can reuse the `struct iovec[]` as the kernel did make a copy of it. That the whole idea of the "stable feature". You cant just reuse the buffers itself.

That said here is the fix https://github.com/netty/netty/pull/10627/

**Comment by 1Jo1** [comment]
@normanmaurer ok I didn't know that, I thought it would just do private of the sqe entries, not from the buffer

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java [hunk 3]
```diff
@@ -342,11 +342,21 @@ private int flushSqe() {
     public void submit() {
         int submitted = flushSqe();
         logger.trace("Submitted: {}", submitted);
-        if (submitted > 0) {
+        if (submitted <= 0) {
+            return;
+        }
+        if (!Native.SQPOLL) {
```
Comment (normanmaurer) [inline transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java:348]: can we just make this a boolean constructor flag per `IOUringEventLoop` ?

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java [hunk 4]
```diff
@@ -342,11 +342,21 @@ private int flushSqe() {
...
             int ret = Native.ioUringEnter(ringFd, submitted, 0, 0);
             if (ret < 0) {
                 throw new RuntimeException("ioUringEnter syscall");
             }
+        } else {
+
```
Comment (normanmaurer) [inline transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java:354]: nit: `else if ...` (you can merge the two lines

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java [hunk 5]
```diff
@@ -342,11 +342,21 @@ private int flushSqe() {
...
+            //release barrier kFlagAddress
+            if ((PlatformDependent.getIntVolatile(kFlagsAddress) & Native.IORING_NEED_WAKEUP) != 0) {
+                //System.out.println("Wakeup Thread");
```
Comment (normanmaurer) [inline transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java:357]: nit: remove

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/Native.java [hunk 6]
```diff
@@ -33,6 +33,9 @@
     // Todo expose this via the EventLoopGroup constructor as well.
     private static final int DEFAULT_RING_SIZE = SystemPropertyUtil.getInt("io.netty.uring.ringSize", 4096);
 
+    // Todo configurable
+    static final boolean SQPOLL = true;
```
Comment (normanmaurer) [inline transport-native-io_uring/src/main/java/io/netty/channel/uring/Native.java:37]: See above

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: transport-native-io_uring/src/main/c/netty_io_uring_native.c [hunk 7]
```diff
@@ -400,6 +401,18 @@ static jint netty_io_uring_iosqeAsync(JNIEnv* env, jclass clazz) {
     return IOSQE_ASYNC;
 }
 
+static jint netty_io_uring_sqPoll(JNIEnv* env, jclass clazz) {
```
Comment (normanmaurer) [inline transport-native-io_uring/src/main/c/netty_io_uring_native.c:404]: nit... following our style we would call it `netty_io_uring_setup_sqpoll`

File: transport-native-io_uring/src/main/c/netty_io_uring_native.c [hunk 8]
```diff
@@ -400,6 +401,18 @@ static jint netty_io_uring_iosqeAsync(JNIEnv* env, jclass clazz) {
...
+    return IORING_SETUP_SQPOLL;
+}
+
+static jint netty_io_uring_sqNeedWakeUp(JNIEnv* env, jclass clazz) {
+    return IORING_SQ_NEED_WAKEUP;
+}
+
+static jint netty_io_uring_enter_wakeup(JNIEnv *env, jclass clazz) {
```
Comment (normanmaurer) [inline transport-native-io_uring/src/main/c/netty_io_uring_native.c:412]: nit... following our style we would call it `netty_io_uring_enter_wakup`

**Comment by normanmaurer** [comment]
@1Jo1 can you make the requested changes or should I pick it up and push it over the finish line ?

**Comment by 1Jo1** [comment]
@normanmaurer I'm working on it, still rebasing...

**Comment by normanmaurer** [comment]
@1Jo1 all the fun... :D

**Comment by 1Jo1** [comment]
SQPOLL non-fixed file support is deferred it to 5.11

**Comment by normanmaurer** [comment]
Please re-open in https://github.com/netty/netty-incubator-transport-io_uring which is the new home of the io-uring transport

**Pull request closed without merging**